flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java

```diff
@@ -906,11 +907,33 @@ private void resetAvailableTransactionalIdsPool(Collection<String> transactional
 	// ----------------------------------- Utilities --------------------------
 
 	private void abortTransactions(Set<String> transactionalIds) {
-		for (String transactionalId : transactionalIds) {
-			try (FlinkKafkaProducer<byte[], byte[]> kafkaProducer =
-					initTransactionalProducer(transactionalId, false)) {
-				// it suffice to call initTransactions - this will abort any lingering transactions
-				kafkaProducer.initTransactions();
+		// shortcut for non-exactly-once producers
+		if (transactionalIds.isEmpty()) {
+			return;
+		}
+
+		ForkJoinPool forkJoinPool = null;
+		try {
+			// limit the number of connections to the number that is used during runtime
+			forkJoinPool = new ForkJoinPool(kafkaProducersPoolSize);
+			forkJoinPool.submit(() ->
+				transactionalIds.parallelStream().forEach(transactionalId -> {
+					// don't mess with the original configuration or any other properties of the
+					// original object
+					// -> create an internal kafka producer on our own and do not rely on
+					//    initTransactionalProducer().
+					final Properties myConfig = new Properties();
+					myConfig.putAll(producerConfig);
+					myConfig.put("transactional.id", transactionalId);
```

Comment:
can you deduplicate the config setting code with `initTransactionalProducer`? Yes, now it's only one `put("transactional.id", transactionalId);` call, but this can change at some point of time and such code duplication can make this fragile.